CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+#include "DesktopNotificationManagerCompat.h"
+#include <appmodel.h>
+#include <string>
+
+using namespace Microsoft::WRL;
+using namespace Microsoft::WRL::Wrappers;
+
+bool DesktopNotificationManagerCompat::s_registeredAumidAndComServer = false;
+const wchar_t *DesktopNotificationManagerCompat::s_aumid = nullptr;
+bool DesktopNotificationManagerCompat::s_registeredActivator = false;
+bool DesktopNotificationManagerCompat::s_hasCheckedIsRunningAsUwp = false;
+bool DesktopNotificationManagerCompat::s_isRunningAsUwp = false;
+
+HRESULT DesktopNotificationManagerCompat::RegisterAumidAndComServer(const wchar_t *aumid, GUID clsid)
+{
+    // If running as Desktop Bridge
+    if (IsRunningAsUwp())
+    {
+        // Clear the AUMID since Desktop Bridge doesn't use it, and then we're done.
+        // Desktop Bridge apps are registered with platform through their manifest.
+        // Their LocalServer32 key is also registered through their manifest.
+        s_aumid = nullptr;
+        s_registeredAumidAndComServer = true;
+        return S_OK;
+    }
+
+    s_aumid = aumid;
+
+    // Get the EXE path
+    wchar_t exePath[MAX_PATH];
+    DWORD charWritten = ::GetModuleFileName(nullptr, exePath, ARRAYSIZE(exePath));
+    auto hr = charWritten > 0 ? S_OK : HRESULT_FROM_WIN32(::GetLastError());
+    if (SUCCEEDED(hr))
+    {
+        // Register the COM server
+        hr = RegisterComServer(clsid, exePath);
+        if (SUCCEEDED(hr))
+        {
+            s_registeredAumidAndComServer = true;
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterActivator()
+{
+    // Module<OutOfProc> needs a callback registered before it can be used.
+    // Since we don't care about when it shuts down, we'll pass an empty lambda here.
+    Module<OutOfProc>::Create([] {});
+
+    // If a local server process only hosts the COM object then COM expects
+    // the COM server host to shutdown when the references drop to zero.
+    // Since the user might still be using the program after activating the notification,
+    // we don't want to shutdown immediately.  Incrementing the object count tells COM that
+    // we aren't done yet.
+    Module<OutOfProc>::GetModule().IncrementObjectCount();
+
+    auto hr = Module<OutOfProc>::GetModule().RegisterObjects();
+    if (SUCCEEDED(hr))
+    {
+        s_registeredActivator = true;
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterComServer(GUID clsid, wchar_t exePath[])
+{
+    // Turn the GUID into a string
+    OLECHAR* clsidOlechar;
+    StringFromCLSID(clsid, &clsidOlechar);
+    std::wstring clsidStr(clsidOlechar);
```

Comment:
This can throw. We shouldn't mix throw and HRESULT. Prefer throwing so that you can use std, but if the code wants HRESULT then everything should be exception safe.

Reply:
Ah nice catch. As for HRESULTs vs throwing - what's the preferred pattern in C++? Our UWP APIs seem to return HRESULTS, but if most prefer throwing, we could make the library throw instead. I just don't know what C++ devs prefer :)

Reply:
throwing is preferred. Our WinRT APIs return HRESULTs because c++ exceptions don't translate across the ABI correctly. Your code is compiled inside their app and that won't be an issue.

Reply:
Awesome! I assume that means I should make any failed HRESULTS also throw, and therefore make a lot of these methods void (and make some of them return objects rather than set the pointer object in a parameter)?